eli5/keras.py

```diff
@@ -0,0 +1,293 @@
+    # Some ideas: 
+    # check layer type, i.e.: isinstance(l, keras.layers.Conv2D)
+    # check layer name
+    l = estimator.get_layer(index=i)
+    # a check that asks 'can we resize this activation layer over the image?'
+    rank = len(l.output_shape)
+    required_rank = len(estimator.input_shape)
+    return rank == required_rank
+
+
+def get_target_prediction(model, x, targets):
+    """
+    Get a prediction ID from ``targets``.
+
+    See documentation of ``explain_prediction_keras`` for explanation of ``targets``.
+    
+    Returns
+    -------
+    prediction id : int
+    """
+    # TODO: take in a single target as well, not just a list, 
+    # consider changing signature / types for explain_prediction generic function
+    # TODO: need to find a way to show the label for the passed prediction 
+    # as well as its probability
+
+    # TODO: maybe do the sum / loss in this function instead of grad_cam. Return a tensor.
+    # This would be consistent with what is done in https://github.com/ramprs/grad-cam/blob/master/misc/utils.lua
+    # https://github.com/ramprs/grad-cam/blob/master/classification.lua
+    # https://github.com/torch/nn/blob/master/doc/module.md
+    if isinstance(targets, list):
+        # take the first prediction from the list
+        predicted_idx = targets[0]
+        # TODO: use all predictions in the list
+        # TODO: validate list contents
+    elif targets is None:
+        predictions = model.predict(x)
+        predicted_idx = np.argmax(predictions)
+    else:
+        raise ValueError('Invalid argument "targets" (must be list or None): %s' % targets)
+    return predicted_idx
+
+
+def grad_cam(estimator, image, prediction_index, activation_layer):
+    """
+    Generate a heatmap using Gradient-weighted Class Activation Mapping (Grad-CAM).
+
+    Credits for implementation
+    * Jacob Gildenblat for "https://github.com/jacobgil/keras-grad-cam".
+    * Author of "https://github.com/PowerOfCreation/keras-grad-cam" for fixes to Jacob's implementation.
+    * Kotikalapudi, Raghavendra and contributors for "https://github.com/raghakot/keras-vis".
+    
+    Returns
+    -------
+    heatmap : object
+        A numpy.ndarray localization map.
+    """
+    # FIXME: this assumes that we are doing classification
+    # FIXME: we also explicitly assume that we are dealing with images
+
+    # Get required terms
+    weights, activations, grads_val = grad_cam_backend(estimator, image, prediction_index, activation_layer)
+
+    # Perform a weighted linear combination
+    spatial_shape = activations.shape[:2]
+    lmap = np.zeros(spatial_shape, dtype=np.float32)
+    for i, w in enumerate(weights):
+        # weight * single activation map
+        # add to the entire map (linear combination), NOT pixel by pixel
+        lmap += w * activations[..., i]
+
+    lmap = np.maximum(lmap, 0) # ReLU
+
+    lmap = lmap / np.max(lmap) # -> [0, 1] ndarray
```

Comment:
If ``np.max(lmap)`` can be zero, would be nice to add some eps, e.g. ``lmap / (eps + np.max(lmap))``, or maybe there is already an existing function for that.

Reply:
https://github.com/TeamHG-Memex/eli5/pull/315/commits/9151d4ca5c30656fd8305185935769025ff67575